internal/server/data/provideruser.go

```diff
@@ -47,26 +70,88 @@ func CreateProviderUser(db GormTxn, provider *models.Provider, ident *models.Ide
 	if err := validateProviderUser(pu); err != nil {
 		return nil, err
 	}
-	return pu, add(db, pu)
+
+	if err := insert(db, (*providerUserTable)(pu)); err != nil {
+		return nil, err
+	}
+	return pu, nil
 }
 
-func UpdateProviderUser(db GormTxn, providerUser *models.ProviderUser) error {
+func UpdateProviderUser(tx WriteTxn, providerUser *models.ProviderUser) error {
 	if err := validateProviderUser(providerUser); err != nil {
 		return err
 	}
-	return save(db, providerUser)
+	providerUser.LastUpdate = time.Now().UTC()
+
+	pu := (*providerUserTable)(providerUser)
+	query := querybuilder.New("UPDATE")
+	query.B(pu.Table())
+	query.B("SET")
+	query.B(columnsForUpdate(pu), pu.Values()...)
+	query.B("WHERE provider_id = ? AND identity_id = ?;", providerUser.ProviderID, providerUser.IdentityID)
+	_, err := tx.Exec(query.String(), query.Args...)
+	return handleError(err)
 }
 
-func ListProviderUsers(db GormTxn, p *Pagination, selectors ...SelectorFunc) ([]models.ProviderUser, error) {
-	return list[models.ProviderUser](db, p, selectors...)
+func ListProviderUsers(tx ReadTxn, providerID uid.ID) ([]models.ProviderUser, error) {
+	table := &providerUserTable{}
+	query := querybuilder.New("SELECT")
+	query.B(columnsForSelect(table))
+	query.B("FROM")
+	query.B(table.Table())
+	query.B("WHERE provider_id = ?", providerID)
```

Comment:
While reading over these queries it is now a lot more obvious that we always use `provider_id` but not always `identity_id`. I opened #3347 to fix the primary key index so that we can make use of it for all of these queries.